### DISCUSSION TITLE
当下游服务返回"text/plain"类型时导致"response.Body"中变得异常得长,这正常吗?

### DISCUSSION THREAD

**Discussion by 17371** [ThreeMammals/Ocelot#2047] [Q&A]
## 我期望能将响应内容进行存储,但意外发现响应内容无法解析,他将显示乱码; 调试发现response.ContentType:application/json 没问题,当response.ContentType为text/plain时出现问题
###下面是我的描述
例如下游服务将返回
> ServerA Program.cs:

 ```
app.UseMiddleware<MyCustomMiddleware>();
...
app.MapGet("/test", () =>  " 运行正常");
```
﻿

> OcelotGateway Program.cs:
```
app.UseMiddleware<MyCustomMiddleware>();
app.UseOcelot().Wait();
```
>MyCustomMiddleware
```
 Stream originalResponseStream = response.Body;
 using MemoryStream replaceResponseStream = new MemoryStream();
 response.Body = replaceResponseStream;
 await next(context); 
 res_content_type = response.ContentType ?? "";
 Console.WriteLine("响应类型:" + res_content_type); 
 Console.WriteLine("响应长度:" + replaceResponseStream.Length);
 replaceResponseStream.Position = 0;
 await replaceResponseStream.CopyToAsync(originalResponseStream);
 response.Body = originalResponseStream;
```
## 
> ServerA 
响应类型:text/plain; charset=utf-8
﻿响应长度:2

> ﻿OcelotGateway 
﻿﻿响应类型:text/plain; charset=utf-8
﻿响应长度:121
﻿
### replaceResponseStream从原来得2变成了121 !!!我不是很理解多出得部分是什么
﻿
﻿

**Comment by raman-m** [comment]
@ggnaegi What do you think?
It will be a surprise for me if gateway can't handle `text/plain` **ContentType** header requests correctly.
Please find some time to answer to the author.

**Reply by raman-m** [reply]
@ggnaegi Pay your attention to this `Should_add_default_content_type_header` test please!
https://github.com/ThreeMammals/Ocelot/blob/59b63eab9c2cd2ab33c5752bbc3bd00a079cd74c/test/Ocelot.AcceptanceTests/ContentTests.cs#L73

**Comment by 17371** [comment]
Sorry for taking up your time!
I think the content of response.Body was changed in app.UseOcelot(), but I don’t know where the change was made. What’s strange to me is that text/plain was changed separately, but application/json was not affected. . Maybe I should read out the bytes and compare.

**Reply by raman-m** [reply]
### First
From issue title: 
- **There was a problem when ContentType was text/plain**
- I accidentally found that the content of the response could not be parsed, and it displays garbled characters;

So, it could be text stream encoding issue...
What's the value of `Content-Encoding` header?
What's the value of `Accept-Encoding` header?
If you have something different than `utf-8` then you could see incorrectly displayed text! 
See more in #1072 and #1900 

### Second

> > ServerA Program.cs:
> 
>  ```
> app.UseMiddleware<MyCustomMiddleware>();
> ...
> app.MapGet("/test", () =>  " 运行正常");
> ```
> ﻿
> 
> > OcelotGateway Program.cs:
> ```
> app.UseMiddleware<MyCustomMiddleware>(); // Please don't do it for Ocelot pipeline! You must be careful!!!
> app.UseOcelot().Wait();
> ```

You did some coding research and you used custom middleware which is not official Ocelot's one. So you've changed Ocelot pipeline ❗ 
We don't accept issues of changed Ocelot pipeline! And I don't see any benefits to continue this discussion!
Going to close the discussion...

**Reply by 17371** [reply]
OK, thanks

**Comment by 17371** [comment]
OK, thanks
